### PR TITLE
Fix “No such file or directory” update_check.json

### PR DESCRIPTION
**Pull request by zonuexe**
```
% php ./my-shell.php
Psy Shell v0.8.0 (PHP 5.5.26 — cli) by Justin Hileman
ErrorException with message 'file_get_contents(/home/tadsan/.config/psysh/update_check.json): failed to open stream: No such file or directory'
Bye...
```

### CODE REVIEW

File: src/Psy/Configuration.php [hunk 1]
```diff
@@ -1110,7 +1110,12 @@ public function getChecker()
                 case Checker::DAILY:
                 case Checker::WEEKLY:
                 case Checker::MONTHLY:
-                    $this->checker = new IntervalChecker($this->getUpdateCheckCacheFile(), $interval);
+                    $check_file = $this->getUpdateCheckCacheFile();
+                    if ($check_file === false) {
+                        $this->checker = new NoopChecker();
```
Comment (bobthecow) [inline src/Psy/Configuration.php]: i think falling back to a `NoopChecker` when the user hasn't explicitly turned off update checking is probably grounds for showing an error message or warning of some sort.

File: src/Psy/Configuration.php [hunk 2]
```diff
@@ -1162,15 +1167,30 @@ public function setUpdateCheck($interval)
...
      * Get a cache file path for the update checker.
      *
-     * @return string
+     * @return string|false Return false if config file/directory is not writable
      */
     public function getUpdateCheckCacheFile()
     {
         $dir = $this->configDir ?: ConfigPaths::getCurrentConfigDir();
+
+        if (!is_writable($dir)) {
```
Comment (bobthecow) [inline src/Psy/Configuration.php]: this problem needs to be solved for the history file as well. any thoughts on a more general solution?

File: src/Psy/ConfigPaths.php [hunk 3]
```diff
@@ -180,4 +180,41 @@ private static function getRealFiles(array $dirNames, array $fileNames)
...
+     *
+     * @param string $dir
+     * @param string $file
+     *
+     * @return string|false Full path to $file. Return false if file is not writable.
+     */
+    public static function touchFileWithMkdir($dir, $file)
+    {
+        if (!is_writable($dir)) {
+            trigger_error("Writing to {$dir} is not allowed.", E_USER_NOTICE);
```
Comment (bobthecow) [inline src/Psy/ConfigPaths.php]: If `$path` exists and is writable, but `$dir` isn't writable, we probably shouldn't treat that as an error.

File: src/Psy/Configuration.php [hunk 4]
```diff
@@ -1110,7 +1112,12 @@ public function getChecker()
                 case Checker::DAILY:
                 case Checker::WEEKLY:
                 case Checker::MONTHLY:
-                    $this->checker = new IntervalChecker($this->getUpdateCheckCacheFile(), $interval);
+                    $check_file = $this->getUpdateCheckCacheFile();
```
Comment (bobthecow) [inline src/Psy/Configuration.php]: Please use `$checkFile` for consistency.

File: src/Psy/Configuration.php [hunk 5]
```diff
@@ -1162,15 +1169,12 @@ public function setUpdateCheck($interval)
...
      */
     public function getUpdateCheckCacheFile()
     {
         $dir = $this->configDir ?: ConfigPaths::getCurrentConfigDir();
-        if (!is_dir($dir)) {
-            mkdir($dir, 0700, true);
-        }
 
-        return $dir . '/update_check.json';
+        return ConfigPaths::touchFileWithMkdir($dir, '/update_check.json');
```
Comment (bobthecow) [inline src/Psy/Configuration.php]: Leave the slash off the file name?

Comment (zonuexe) [inline src/Psy/Configuration.php]: Fixed.

File: src/Psy/Configuration.php (hunk 4 above)
Comment (zonuexe) [inline src/Psy/Configuration.php]: Fixed.

**Comment by bobthecow** [comment]
The `is_writable` check was in the wrong place, because a directory that doesn't exist by definition isn't writable, which is why CI was failing. But I've merged this PR along with a fix. Thanks!

**Comment by zonuexe** [comment]
🙇

**Pull request merged**